### ISSUE TITLE
Enum encode takes an Enum, decode returns a string

### ISSUE THREAD

**Issue by kklipsch** [linkedin/goavro#59]
When encoding an enum field on a record it requires a Enum:

https://github.com/linkedin/goavro/blob/master/codec.go#L481

When decoding an enum field it returns a string:

https://github.com/linkedin/goavro/blob/master/codec.go#L478

This is inconsistent.

**Comment by slaunay** [comment]
It looks like this had been introduced by the last commit: d3e5b23 for #38.

By having a non symmetric API, it is not possible to read a datum containing an enum field and write it back:
```
r := os.Open("...")
reader, _ := goavro.NewReader(goavro.BufferFromReader(r))
codec, _ := goavro.NewCodec(reader.DataSchema)
datum, _ := reader.Read();
bb := new(bytes.Buffer)
err := a.codec.Encode(bb, datum) // Will fail here with: cannot encode enum (...): expected: Enum; received: string
```

Returning a `goavro.Enum` during decoding would fix this but it is not a backward compatible change.

**Comment by karrick** [comment]
@slaunay,

That is precisely what happened. The original intention was for goavro to perform as much of the data marshaling as possible, thus reducing the amount of boxing and unboxing of values. The Record type is the biggest exception, and later on we had to create an Enum type.

The reason we had to create the Enum type was to deal with Avro unions. When a schema specified the data type to be a union of two other types, one type of which was a string, and the other an Avro enum, it's easy to know what to do when we decode a datum. However when a program was _encoding_ a datum and received a string, it seemed silly to first check whether that string was one of the enums. If the user had meant to encode a string, but that string was later added to the Avro enum, it would later be encoded as an Enum rather than a string.
